### ISSUE TITLE
Can't build tarballs from codeberg

### ISSUE THREAD

**Issue by andrei-a-papou** [vonshednob/pter#41]
I can't build https://codeberg.org/vonshednob/pter/archive/v3.15.2.tar.gz, the error is: `error: can't copy 'man/pter.1': doesn't exist or not a regular file`

Here's how I'm building:
`python3 ./setup.py build && python3 ./setup.py install`

Same thing happened with https://codeberg.org/vonshednob/pter/archive/v3.15.1.tar.gz, but interestingly I just downloaded https://vonshednob.cc/releases/pter-3.15.1.tar.gz and there was no problem. Apparently, the tarballs at codeberg and https://vonshednob.cc are different.

There's no tarball for 3.15.2 at https://vonshednob.cc, so now I'm stuck :) Can't install via pip as I need to patch the tarball first with my own set of patches.

**Comment by andrei-a-papou** [comment]
OK, seems like `apt install python3-docutils` works around the issue, but shouldn't `docutils` be fetched by setup.py somehow and/or listed as deps? And why doesn't the issue occur with https://vonshednob.cc/releases/?

**Comment by vonshednob** [comment]
The tarball from my website has been produced with `python -m build`, so it’s not just the source code, but contains the generated man page, too.

I was rather reluctant to add docutils as a build-time dependency -- the pter package on pypi and on the website, as well as the one in AUR (all the same file, from `python -m build`) "just work". So adding docutils as a build-time dependency would be an unnecessary step. :thinking:

How about adding a Development section to the README and add that type of information there?

**Comment by andrei-a-papou** [comment]
> How about adding a Development section to the README and add that type of information there?

IMO, that would be fine. For the record, this is what I use to build pter inside a Docker image (based on Debian 12/Bookworm):
```
apt-get install --no-install-recommends -y \
    python3.11-minimal python3-setuptools python3-docutils ca-certificates
```

`ca-certificates` are needed on a minimal Debian install -- without the package no deps like `pytodotxt` can be downloaded.

**Comment by vonshednob** [comment]
Interesting, but I remember from my own experience with … uhm, something, that `ca-certificates` surprisingly is not installed by default. But I guess it makes sense for containers, similar to not installing man pages.

**Comment by andrei-a-papou** [comment]
> But I guess it makes sense for containers, similar to not installing man pages.
Yeah, this is inside the `debian:bookworm-slim` image. I guess they save space by not installing `ca-certificates`.

> How about adding a Development section to the README and add that type of information there?
I've played with several build options recently and yes, adding instructions on how to build from _both kinds of tarballs_ would be much appreciated. 

Right now I can build from both kinds of tarballs using `setup.py`, but it turns out calling it directly is now deprecated (warnings get printed out). 

Building using `pip` from a source directory fails for me: a man page is not found even after installing `docutils`. This is with 3.15.2, codeberg tarball. So I couldn't convert to using pip, unfortunately.

Running `pip3 install pter` works, but then it's hard or very awkward to apply my patches (of which I have more :))

So I'm stuck with using the deprecated `setup.py`, which works across the board -- for now. It's by no means a blocker, just something that will be hopefully addressed in the future.

**Comment by vonshednob** [comment]
I think the current thing to do is run `python -m build` and `python -m install`. Totally not a circular dependency when you need to install the `install` package first :upside_down_face:

**Comment by andrei-a-papou** [comment]
OK, I've tries that, it appears to print the same warnings as pip:

```
warning: no files found matching '*.html' under directory 'pter'                                                                                                                               
warning: no files found matching '*' under directory 'man' 
```

Then errors out:
```
error: can't copy 'man/pter.1': doesn't exist or not a regular file
```

This is with codeberg tarball, v3.15.2. Package `python3-docutils` installed via `apt-get`

**Comment by vonshednob** [comment]
That's odd. With docutils it should be able to build the man page during `python -m build` :thinking: maybe there's another dependency that I forgot

**Comment by vonshednob** [comment]
Okay, I give up for now -- apparently it's far from straight forward to declare a build-time dependency in python. I’ll try to make it not burst into flames when python is being build without having the `docutils` installed: [e12dc7e0](https://codeberg.org/vonshednob/pter/commit/e12dc7e02e51b3c70fcfd3f8f996cd9f3acb0723). Let me know if this worked.

**Comment by andrei-a-papou** [comment]
Oh man, now I'm getting `error: can't copy 'man/pter.1': doesn't exist or not a regular file` even when building from git. 

This happens when running `python3 ./setup.py build && python3 ./setup.py install`

**Comment by vonshednob** [comment]
Curious. I just clean-built in an empty virtualenv with that new configuration and it worked. At what step does it crash for you? `python setup.py build` or `python setup.py install`?

**Comment by vonshednob** [comment]
Could you try to use `pip install .` instead of `python setup.py`?

**Comment by andrei-a-papou** [comment]
I'm sorry, the error is slightly different, it's in regard to qpter, not pter. Sorry for the confusion.

**Comment by vonshednob** [comment]
> I'm sorry, the error is slightly different, it's in regard to qpter, not pter. Sorry for the confusion.

No worries, I also missed that.

> Here's the actual error: `error: can't copy 'man/pter/qpter.1': doesn't exist or not a regular file`

Yeah, that was me fumbling the patch. There really is no `man/pter/qpter.1` file, but a `man/qpter.1` file. :facepalm: 

Anyway, should be fixed with [12ea984a](https://codeberg.org/vonshednob/pter/commit/12ea984a6bed3f291467ec03f5835c3e4e94f64e)

**Comment by andrei-a-papou** [comment]
Now it builds fine via `python3 ./setup.py build && python3 ./setup.py install`.

But there are a bunch of warnings (they don't appear when building via `pip`):

```                                                                                                                              
doc/pter.config.rst:213: (WARNING/2) Inline literal start-string without end-string.                                                                                                           
doc/pter.config.rst:213: (WARNING/2) Inline literal start-string without end-string.                                                                                                           
doc/pter.config.rst:213: (WARNING/2) Inline literal start-string without end-string.                                                                                                           
doc/pter.config.rst:213: (WARNING/2) Inline literal start-string without end-string. 
```

The warnings are for the `build` stage. There are other warnings but they are related to the deprecated use of setup.py.

**Comment by vonshednob** [comment]
The line 213 warning is … well, a speciality of trying to express "`+-@ `" (the last character is a space) in reStructuredText. I’ll verify this later. Thanks for the heads-up!

I guess the big win is that it builds and installs now.

**Comment by andrei-a-papou** [comment]
> I guess the big win is that it builds and installs now.

Absolutely!

**Comment by vonshednob** [comment]
I think I might have fixed the warning. D'you think this issue can be closed?

**Comment by andrei-a-papou** [comment]
Tested with both `setup.py` and with `pip3`, you certainly did! 

> D'you think this issue can be closed?

Sure, it builds just fine now.

**Comment by andrei-a-papou** [comment]
@vonshednob By the way, will you keep uploading tarballs to https://vonshednob.cc/pter/releases.html? I'm asking since one can build man pages form https://vonshednob.cc releases, which doesn't happen with tarballs from codeberg. And I sometimes need to do `man pter` :)

**Comment by vonshednob** [comment]
Absolutely! Documentation on the terminal is very much required in my opinion. Just haven't released 3.16 yet.
